Add `bundlewatch/bundlewatch` to dev dependencies.

tomalec opened the pull request:
### Changes proposed in this Pull Request:
Add `bundlewatch/bundlewatch` to dev dependencies.
to monitor our progress in reducing bundle size.

Fixes https://github.com/woocommerce/google-listings-and-ads/issues/818.


### Screenshots:
![bundlewatch](https://user-images.githubusercontent.com/17435/123252811-fa33c800-d4ec-11eb-89c3-87d5cdbf22d1.gif)



### Detailed test instructions:

1. check out this branch
2. `npm install`
3. `npx bundlewatch`
4. Click on the link for the detailed report

### To be done:
- [ ] Authorize the app, to enable GitHub check. https://service.bundlewatch.io/setup-github


### Changelog entry

> Tweak - Add bundlewatch to monitor bundle size in GitHub PRs.

tomalec committed:
Add `bundlewatch/bundlewatch` to dev dependencies.

to monitor our progress in reducing bundle size.

Fixes https://github.com/woocommerce/google-listings-and-ads/issues/818.

tomalec commented:
#### Small research for [`bundlewatch` usage in Woo](https://github.com/search?q=org%3Awoocommerce+bundlewatch&type=commits)

`gutenberg-products-block` repo [tried it](https://github.com/woocommerce/woocommerce-gutenberg-products-block/commit/2c01e0f683aca6439dc61e84feeb5ac069b4d400) then [removed it](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/2005) over a year ago in favor of custom GH workflow, for the following reasons:

> - We get inline feedback (as a comment) in the pull on any size changes.

Personally, I find it a disadvantage:
- It comments only the initial PR post (commits set), so we don't get bundle size after commits that may follow during the review. In GLA we tend to make quite a number of commits after the initial PR.
- This way we cannot block the merge if we enforce a bundle size budget

> - This compares overall changes against the base branch for the pull (which we weren't getting with bundlewatch).

If I understood [the docs](https://bundlewatch.io/#/getting-started/diffing-against-branches-other-than-master) correctly, the current version of `bundlewatch` can pick the PR base from Travis and compare against it.

> - A list of all files and their changes is shown in a collapsed list in the posted comment. With bundlewatch, their linking system was broken, so if we wanted that granularity we had to go into the travis-ci job (which only showed us the sizes, not the difference from the base branch).

For my local test, linking was working fine. Also, I think in GLA we do not have so many bundles, so hopefully, it's not a big issue for us.

> - Removes the time for the bundlewatch checks from the linting travis job (not a huge difference but helps).

If we're not experiencing hitting Travis limit yet, maybe we can still use it, and worry about it once we start exceeding it.

----

woocommerce.com [is using `bundlesize`](https://github.com/Automattic/woocommerce.com/blob/trunk/package.json#L20) the ancestor of `bundlewatch`

----

In `woocommerce-admin` some efforts were made to try to setup a custom github action, but it seems it wasn't finished
https://github.com/woocommerce/woocommerce-admin/actions/workflows/bundle-size.yml

tomalec commented:
I think the above gives us precedence to trust the `bundlewatch` enough to authorize it, WDYT @jconroy?

ecgan commented:
@tomalec , I love the analysis you posted, good stuff 🙂 👍 

> woocommerce.com [is using `bundlesize`](https://github.com/Automattic/woocommerce.com/blob/trunk/package.json#L20) the ancestor of `bundlewatch`

That got me curious. I looked into both `bundlesize` and `bundlewatch`, and it seems like `bundlesize` is much more popular than `bundlewatch`. 

- https://github.com/siddharthkp/bundlesize: 4.2k stars, 207 forks, 82k weekly downloads in https://www.npmjs.com/package/bundlesize.
- https://github.com/bundlewatch/bundlewatch: 263 stars, 28 forks, 69k weekly downloads in https://www.npmjs.com/package/bundlewatch.

I tried to google for comparisons between the two to find out which is better, but there isn't any helpful info on this.

So my question here is: given its popularity, should we consider using `bundlesize` instead of `bundlewatch`? Why or why not? 🤔

tomalec commented:
> […] given its popularity, should we consider using `bundlesize` instead of `bundlewatch`? Why or why not? thinking

AFAIK `bundlewatch` is kindof community fork of `bundlesize`, as `-size` got somewhat unmaintained.
You can check [`bundlewatch`'es list of reasons](https://github.com/bundlewatch/bundlewatch#i-use-bundlesize-why-should-i-switch-to-bundlewatch)
> - Bundlesize has entered maintenance mode and pull requests are left hanging, so we wanted to reboot the community through creating BundleWatch


Check the latest release dates:
 - https://github.com/siddharthkp/bundlesize/releases - `0.18.0`  Jun 29, **2019**
 - https://github.com/bundlewatch/bundlewatch/releases - `0.3.2` Jan 26, **2021**

findingsimple commented:
Thanks @tomalec for looking into this and especially [for searching around in this comment](https://github.com/woocommerce/google-listings-and-ads/pull/819#issuecomment-872936838) to look for existing usages within Woo/A8C

> Authorize the app, to enable GitHub check

<img width="812" alt="Screen Shot 2021-07-06 at 9 48 53 pm" src="https://user-images.githubusercontent.com/355014/124598869-3f200d00-dea4-11eb-814d-96d6c89ed344.png">

I think both BundleSize and BundleWatch area already approved for the Woo org.

I like the idea of using Actions (especially now the repo is public) - we'll likely use more - but happy for us to try BundleWatch and revisit it and see if the previous areas of concern have improved.

I leave @ecgan and @eason9487 to comment/feedback/review before merging

ecgan commented:
@tomalec ,

> You can check [`bundlewatch`'es list of reasons](https://github.com/bundlewatch/bundlewatch#i-use-bundlesize-why-should-i-switch-to-bundlewatch)

Hey, thanks for that! I didn't notice that. Google search should have shown me that 😆 

> Check the latest release dates:
> 
> * [siddharthkp/bundlesize/releases](https://github.com/siddharthkp/bundlesize/releases) - `0.18.0`  Jun 29, **2019**

And I did not notice that too. I did notice that on https://www.npmjs.com/package/bundlesize, the latest release is `0.18.1` and it is published 6 months ago. I looked into the repo and the release is related to `axios` security vulnerability issue (https://github.com/siddharthkp/bundlesize/issues/369), and from that issue I notice there is another repo https://github.com/siddharthkp/bundlesize2, which does not seem to be very active either. 

With all the above said (including your good findings), I think I would lean towards bundlewatch now. Just hoping that it can do everything that `bundlesize` does, and there are good documentations that can be easily googled. 😄

ecgan commented on package.json:
```diff
@@ -89,5 +90,30 @@
 			"src"
 		]
+	},
+	"bundlewatch": {
+		"files": [
+			{
+				"path": "./js/build/*.js",
+				"maxSize": "2 kB"
+			},
+			{
+				"path": "./js/build/index.js",
+				"maxSize": "1 mB"
+			},
+			{
+				"path": "./js/build/*.css",
+				"maxSize": "1 kB"
+			},
+			{
+				"path": "./js/build/index.css",
+				"maxSize": "8 kB"
+			},
+			{
+				"path": "./google-listings-and-ads.zip",
+				"maxSize": "12 mB",
+				"compression": "none"
+			}
+		]
```
💅  Personally I would prefer to put these into a separate file like `.bundlewatch.config.js` and have bundlewatch to _automatically_ read the file, unfortunately currently bundlewatch requires us to specify it in the command line like `bundlewatch --config .bundlewatch.config.js` (see the [documentation](https://bundlewatch.io/#/getting-started/using-a-config-file)), which is an unnecessary friction in my opinion. Hopefully they can support that in the future.

So with that said, good to have this `bundlewatch` section here in this `package.json` so that we can just run `npx bundlewatch`. 👍

ecgan reviewed (approved):
Tested as per instruction, working good, LGTM 👍

tomalec commented:
I tried adding GitHub action for bundlewatch, for example below.
I tried to make it run for every push, not just initial PR.
But there is a bug in `bundlewatch` <- `ci-env` https://github.com/bundlewatch/bundlewatch/issues/422
We could hard-code it to compare only against `trunk` and `develop`, but in my opinion comparing against base makes the most sense.

So I [fall back to running it on Travis-CI](https://github.com/woocommerce/google-listings-and-ads/pull/819/commits/ed9cbc72f848a73df2ba24ca84ad6bf4c74bb913), where it works out of the box.

```yml
name: bundlewatch

on: [push]

jobs:
  build:
    if: "!contains(github.event.commits[0].message, '[skip ci]')"
    runs-on: ubuntu-latest
    timeout-minutes: 10

    steps:
    - uses: actions/checkout@v2

    - name: Install npm dependencies
      run: npm ci

    - name: Build
      run: npm run build

    - name: Run bundlewatch
      run: npx bundlewatch
      env:
        BUNDLEWATCH_GITHUB_TOKEN: ${{ secrets.BUNDLEWATCH_GITHUB_TOKEN }}
        CI_COMMIT_SHA: ${{ github.event.pull_request.head.sha }}
        CI_BRANCH_BASE: ${{ github.base_ref || 'trunk' }}
```

tomalec committed:
Merge branch 'trunk' into add/818-bundlewatch

tomalec committed:
Add bundlewatch to .travis.yml

GitHub Actions are a bit too problematic, see
https://github.com/woocommerce/google-listings-and-ads/pull/819#issuecomment-875930738

Remove redundant `npm ci` from `Coding standard check`.
Increase maxSizes of bundles.

ecgan commented on package.json:
```diff
@@ -95,25 +95,28 @@
 			}
-		]
+		],
+        "ci": {
+            "trackBranches": ["trunk", "develop"]
+        }
```
@tomalec , just noticed this, really minor nitpick: the three lines here are using space indentation instead of tabs like other lines. 😄

tomalec replied:
I'd address it in https://github.com/woocommerce/google-listings-and-ads/pull/869
